File: answers/lintcode/Flatten Binary Tree to Linked List.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    void flatten(TreeNode *root) {
        // write your code here
        if (NULL == root) {
            return;
        }
        flatten(root->left);
        flatten(root->right);
        TreeNode *right = root->right;
        root->right = root->left;
        TreeNode *left = root;
        while (left && left->right) {
            left = left->right;
        }
        left->right = right;
        root->left = NULL;
    }
};

File: answers/other/in sap/Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
	def flatten(self, root):
		"""
		:type root: TreeNode
		:rtype: void Do not return anything, modify root in-place instead.
		"""
		
		self.helper(root)

	def helper(self, root):
		end = None
		if root:
			right_begin, right_end = self.helper(root.right)
			if right_end:
				end = right_end
			if root.left:
				left_begin, left_end = self.helper(root.left)
				root.left = None
				root.right = left_begin
				left_end.right = right_begin
				if None == end:
					end = left_end
		if None == end:
			end = root
		return [root, end]